Connector example

zeerorg opened the pull request:
This PR updates the connector example with swarm and kubernetes deployment files.

How it is tested:
1. Clone this Pull request locally.
2. Go into the directory: `cd ./connector-sdk/cmd/tester/yaml`
3. For openfaas deployed on docker swarm do: `docker stack deploy func -c ./docker-compose.yml`
4. For openfaas deployed on kubernetes do: `kubectl create -f ./kubernetes --namespace openfaas`

Run this command to deploy a sample function and see `trigger-func` invocation count growing in ui.

```bash
faas-cli deploy --image=functions/nodeinfo --name=trigger-func --annotation topic=faas-request
```

Signed-off-by: Rishabh Gupta <r.g.gupta@outlook.com>

alexellis commented on cmd/tester/main.go:
```diff
@@ -6,27 +6,56 @@ package main
+	}
+}
+
+func getControllerConfig() *types.ControllerConfig {
```
If running in a the default is `gateway.openfaas:8080` / `gateway:8080` on Swarm. We should just not default at all if not passed and return an error `(types.ControllerConfig, error)`

alexellis commented on cmd/tester/main.go:
```diff
@@ -6,27 +6,56 @@ package main
+	if !ok {
+		gURL = "http://127.0.0.1:8080/"
 	}
+	fmt.Println("Gateway Url: ", gURL)
```
Let's not print this here.

alexellis commented on cmd/tester/main.go:
```diff
@@ -6,27 +6,56 @@ package main
+
 	// Simulate events emitting from queue/pub-sub
 	for {
 		time.Sleep(2 * time.Second)
```
Make this configurable to please.

zeerorg replied:
Should I add an env-var to make this configurable ? Or a simple variable is enough ?

alexellis commented on cmd/tester/main.go:
```diff
@@ -6,27 +6,56 @@ package main
+	config := getControllerConfig()
 
 	controller := types.NewController(creds, config)
 	fmt.Println(controller)
```
Remove this line.

alexellis replied:
Replace `fmt` with `log` for use in the container.

alexellis commented on cmd/tester/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.9.2 as builder
```
Copy the example from the vcenter connector please it has an updated Golang version.

alexellis commented on cmd/tester/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+# Stripping via -ldflags "-s -w" 
+RUN CGO_ENABLED=0 GOOS=linux go build -a -ldflags "-s -w" -installsuffix cgo -o ./connector
+
+FROM alpine
```
Copy the example from the vcenter connector please it has an updated alpine version, we must always pin to `3.8` or whatever version we are targeting.

alexellis commented on cmd/tester/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.9.2 as builder
+RUN mkdir -p /go/src/github.com/openfaas-incubator/connector
+WORKDIR /go/src/github.com/openfaas-incubator/connector
```
The path I guess is `connector-sdk` right? To match the repo.

alexellis replied:
Actually I think it's:

`/go/src/github.com/openfaas-incubator/connector-sdk/cmd/tester/` correct?

alexellis commented on cmd/tester/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+COPY main.go    .
+
+RUN go get -d -v ./...
```
Please remove `go get`

It should use vendoring via `dep` or the correct path.

alexellis commented on cmd/tester/README.md:
```diff
@@ -0,0 +1,18 @@
+
+1. Clone this repository: `git clone https://github.com/openfaas-incubator/connector-sdk.git`
+2. Go into the directory: `cd ./connector-sdk/cmd/tester/yaml`
+3. For openfaas deployed on docker swarm do: `docker stack deploy func -c ./docker-compose.yml`
```
`openfaas` -> OpenFaaS

alexellis replied:
`docker swarm` -> Docker Swarm

zeerorg commented:
I've worked through all the suggestions. This should be good to go.

Some major changes:
1. I removed fetching topics from environment variable (made the example hard to follow). now a single topic is defined.
2. Also removed print_response environment variable was not doing anything

alexellis commented on cmd/tester/main.go:
```diff
@@ -5,28 +5,44 @@
+}
+
+func getControllerConfig() (*types.ControllerConfig, error) {
+	gURL, ok := os.LookupEnv("gateway_url")
```
Change to `gatewayURL`

alexellis commented on cmd/tester/yaml/kubernetes/connector-dep.yml:
```diff
@@ -0,0 +1,35 @@
+      containers:
+      - name: sample-connector
+        image: zeerorg/connector-sample:1.0
+        env:
```
Please add topics as an env-var, this is how connectors are intended to work.

alexellis commented:
The scope of this PR was simply to provide a Dockerfile and Kubernetes YAML, not to change the example behaviour or functionality. Please can you revert the below?

> Some major changes:
> I removed fetching topics from environment variable (made the example hard to follow). now a single topic is defined.
> Also removed print_response environment variable was not doing anything

I think the extracting the configuration loading is fine to keep in the changes, but in a separate PR.

Alex
